controllers/controllers.go

```diff
@@ -18,10 +18,53 @@ package controllers
 
 import (
 	"errors"
+	"time"
+
+	ctrl "sigs.k8s.io/controller-runtime"
 
 	"github.com/openshift-kni/numaresources-operator/pkg/status"
 )
 
+type reconcileStep struct {
+	Result        ctrl.Result
+	ConditionInfo conditionInfo
+	Error         error
+}
+
+// Done returns true if a reconciliation step is fully completed,
+// meaning all the substeps have been performed successfully, false otherwise
+func (rs reconcileStep) Done() bool {
+	return rs.ConditionInfo.Type == status.ConditionAvailable
+}
+
+// EarlyStop returns true if we should shortcut after a reconcile substep,
+// false otherwise
+func (rs reconcileStep) EarlyStop() bool {
+	return rs.ConditionInfo.Type != status.ConditionAvailable
+}
+
+func reconcileStepSuccess() reconcileStep {
```

Comment:
Since we're kinda inventing new abstraction for the different reconciliation I think we should add comments explaining when each step returned.

Reply:
done!